### PR TITLE
feat: implement codegen

### PR DESCRIPTION
**Pull request by bitwalker**
*NOTE:* This PR is large, you aren't expected to review everything here at this time. Instead, I would like to get this merged and we can work out issues, clean up, and refactoring moving forward. From now on I would like all PRs in the backend to be small and focused, my hope is that this is the last big PR we'll have to merge for quite some time.

This contains the remaining codegen work I started some time ago, and will allow us to start doing testing, as well as expand its functionality further, implement optimizations, etc. There is a lot here, so feel free to ask questions if you have them, but I'm happy to pair up with anyone on the team to talk through the code in more detail - it can be tough to do so in this format.

There is still one bug I've identified that I should have fixed Monday, but wanted to get this up before continuing on that. I would also like to note that there are a few modules which are overly large and complex, which I would like to break up into submodules, and just generally clean up a lot of functions in the `Stackify` pass in particular. There was a lot of furious hacking things together here, and I haven't managed to clean up all the cruft from that period just yet, but I plan to do that in subsequent PRs in a more focused fashion.

### CODE REVIEW

**Review by jjcnn — COMMENTED** [review]
First part of my review, covering 
- codegen/masm/intrinsics/*
- codegen/masm/emulators.rs

Superficially the code looks fine. I would suggest that you open issues for the TODOs, known bugs and planned refactorings that exist in the code. It'll be difficult to keep track of outstanding tasks otherwise.

I also worry about the amount of testing that will be required for this. There is _a lot_ of code that needs testing, and I know from experience how tedious it is to write extensive tests, so there is a definite risk that we'll end up with an inadequate test suite simply because we'll stop writing tests when we're bored with it, rather than when there is a sufficient number of tests to ensure a high degree of confidence in the code's correctness.

File: codegen/masm/src/emulator.rs [hunk 1]
```diff
@@ -0,0 +1,1605 @@
...
+        })
+    }
+
+    /// Load `program` into this emulator
+    pub fn load_program(&mut self, program: Program) -> Result<(), EmulationError> {
+        for module in program.modules.into_iter() {
+            self.load_module(module)?;
+        }
+
+        // TODO: Load data segments
```
[the remark this answers is not in the document]
Comment (bitwalker) [reply codegen/masm/src/emulator.rs:378]: No this is for static read-only memory (e.g. constant data) associated with a program. The exact mechanics of how these will be implemented in the VM is still up in the air.

**Review by jjcnn — COMMENTED** [review]
Leaving another partial review so that I can keep track of how far I've gotten.

I've looked through all of `codegen/masm/*`, and I have nothing new to add. There are some things that ought to be refactored, but it's of a similar nature to the stuff I mentioned in my previous partial review.

I'll continue my review tonight.

**Review by jjcnn — APPROVED** [review]
Done.

Nothing immediately stands out other than the things I've already mentioned, but this was a superficial review so I didn't look into things in any great detail.

**Pull request merged**